Fix segfaults during CDX read

the-blank-x opened the pull request:
Fixes #23

the-blank-x committed:
Fix possible null pointer write or memory corruption during CDX read

the-blank-x committed:
Fix null pointer read while parsing CDX file

the-blank-x committed:
Initialize WARC dedup table when CDX dedup file is being read

the-blank-x commented:
This writes invalid WARC-Refers-To-Date headers--oops.

This is because the code would now read the timestamps from the CDX file (which appears to be formatted as `YYYYMMDDhhmmss`, but it is not defined in [the legend], nor is it defined in the specifications from [2006] and [2015]), and write it in the WARC-Refers-To-Date header without rewriting it to conform to the [WARC specs] (which specifies that it is "a UTC timestamp formatted according to [W3CDTF]", i.e. in the form of `YYYY-MM-DDThh:mm:ssZ`).

I can have the code rewrite the CDX timestamp into the UTC timestamp according to the W3CDTF, but I am slightly hesitant to (unless given the okay) since I don't know if all properly formed CDX files will have their timestamps be in the form of `YYYYMMDDhhmmss`.

[the legend]: https://archive.org/web/researcher/cdx_legend.php
[2006]: https://iipc.github.io/warc-specifications/specifications/cdx-format/cdx-2006/
[2015]: https://iipc.github.io/warc-specifications/specifications/cdx-format/cdx-2015/
[WARC specs]: https://iipc.github.io/warc-specifications/specifications/warc-format/warc-1.1/#warc-refers-to-date
[W3CDTF]: https://www.w3.org/TR/NOTE-datetime

the-blank-x commented:
The invalid WARC-Refers-To-Date headers issue should be fixed now

Arkiver2 commented:
Thank you! It looks like ```warc_date = cdx_to_warc_timestamp(date);``` is not `xfree`d after use.

I will need to check this PR closer before merging it.

the-blank-x committed:
Fix memory leaks during CDX read

the-blank-x commented:
Oopsie

(and you're welcome!)

the-blank-x committed:
Initialize WARC dedup table if a WARC is created

If this is not done, wget segfaults if --warc-dedup is not passed (but
--warc-file and --warc-cdx is)
